pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/delegate.go

```diff
@@ -0,0 +1,228 @@
+	}
+
+	if isValidUserToken(token, ctxUserName) {
+		return nil, errors.NewNotFound(oauth.Resource("useroauthaccesstokens"), name)
+	}
+
+	return (*oauthapi.UserOAuthAccessToken)(token), nil
+}
+
+func (r *REST) Watch(ctx context.Context, options *metainternal.ListOptions) (watch.Interface, error) {
+	ctxUserName, err := getUserFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	sanitizedListOpts, err := userFieldSelectorToListOptions(options, ctxUserName)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenListWatcher, err := r.accessTokenStorage.Watch(ctx, sanitizedListOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	watcher := newOAuthAccessTokenWatcher(tokenListWatcher, ctxUserName)
+	go watcher.Watch(ctx)
+	return watcher, nil
+}
+
+func (r *REST) ConvertToTable(ctx context.Context, object runtime.Object, tableOptions runtime.Object) (*metav1.Table, error) {
+	return r.tableConvertor.ConvertToTable(ctx, object, tableOptions)
+}
+
+func (r *REST) NamespaceScoped() bool {
+	return false
+}
+
+func (r *REST) Delete(ctx context.Context, name string, validateFunc rest.ValidateObjectFunc, options *metav1.DeleteOptions) (runtime.Object, bool, error) {
+	ctxUserName, err := getUserFromContext(ctx)
+	if err != nil {
+		return &metav1.Status{Status: metav1.StatusFailure}, false, errors.NewNotFound(oauth.Resource("useroauthaccesstokens"), name)
+	}
+
+	var deletedRV string
+	if options.Preconditions != nil && options.Preconditions.ResourceVersion != nil {
+		deletedRV = *options.Preconditions.ResourceVersion
+	}
+
+	deletedObj, err := r.Get(ctx, name, &metav1.GetOptions{TypeMeta: options.TypeMeta, ResourceVersion: deletedRV})
+	if err != nil {
+		return &metav1.Status{Status: metav1.StatusFailure}, false, err
+	}
+
+	deletedTyped := deletedObj.(*oauthapi.UserOAuthAccessToken)
+	if !isValidUserToken((*oauthapi.OAuthAccessToken)(deletedTyped), ctxUserName) {
+		return &metav1.Status{Status: metav1.StatusFailure}, false, errors.NewNotFound(oauth.Resource("useroauthaccesstokens"), name)
+	}
+
+	// don't mutate the original options
+	newOpts := options.DeepCopy()
+
+	switch {
+	case newOpts.Preconditions == nil:
+		newOpts.Preconditions = &metav1.Preconditions{UID: &deletedTyped.UID}
+
+	case newOpts.Preconditions != nil && newOpts.Preconditions.UID != nil && *newOpts.Preconditions.UID != deletedTyped.UID:
+		// the UID seeked to be deleted no longer exists/never existed
+		return &metav1.Status{Status: metav1.StatusFailure}, false, errors.NewNotFound(oauth.Resource("useroauthaccesstokens"), name)
+
+	default:
+		// Make sure the UID of the just-checked object matches the object we're going to remove
+		newOpts.Preconditions.UID = &deletedTyped.UID
+	}
+
+	return r.accessTokenStorage.Delete(ctx, name, validateFunc, newOpts)
+}
+
+func getUserFromContext(ctx context.Context) (string, error) {
+	userInfo, exists := apirequest.UserFrom(ctx)
+	if !exists {
+		klog.V(4).Infof("no user found in the context")
```

Comment:
do we want this? When can it happen?

Reply:
I am not sure whether it's at all possible, it felt like a safe precaution.

Reply:
Would remove the log output here.

Reply:
or even turn this into `string, bool, error` and return the NotFound in the call sites.